File: backend/src/workflow/attempt/plan_dag.py
"""Plan DAG helpers for one attempt — the planner-authored generator+reducer graph.

Named for the *plan* (not every task is in it: planner/advisor/explorer are
off-spine). ``ordered_plan_tasks`` validates the combined generator+reducer DAG
and enforces the structural rules that keep "every attempt has an exit AND
all work is judged" by construction: **≥1 reducer**, and a lane shape where
every generator feeds another generator or a terminal reducer, and every reducer
directly gates one or more generators.
"""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass
from typing import Protocol, TypeVar

from workflow._core.persistence import TaskRow
from workflow._core.primitives import TaskCenterInvariantViolation
from task import (
    TERMINAL_GENERATOR_STATUSES,
    TaskStatus,
)


class PlanNode(Protocol):
    local_id: str
    needs: tuple[str, ...]


GenNode = TypeVar("GenNode", bound=PlanNode)
RedNode = TypeVar("RedNode", bound=PlanNode)


def ordered_plan_tasks(
    generators: tuple[GenNode, ...],
    reducers: tuple[RedNode, ...],
) -> tuple[tuple[GenNode, ...], tuple[RedNode, ...]]:
    """Validate the combined plan DAG and return both tuples in topo order.

    Raises :class:`TaskCenterInvariantViolation` on a duplicate local id, an
    unknown ``needs`` target, a dependency cycle, an empty reducer set, a
    reducer dependency edge, an empty reducer ``needs`` set, or a dangling
    generator no downstream task needs.
    """
    by_needs: dict[str, tuple[str, ...]] = {}
    duplicates: list[str] = []
    for task in (*generators, *reducers):
        if task.local_id in by_needs:
            duplicates.append(task.local_id)
        else:
            by_needs[task.local_id] = task.needs
    if duplicates:
        raise TaskCenterInvariantViolation(
            f"Plan contains duplicate local ids: {tuple(duplicates)!r}"
        )

    for local_id, needs in by_needs.items():
        missing = [dep for dep in needs if dep not in by_needs]
        if missing:
            raise TaskCenterInvariantViolation(
                f"Plan task {local_id!r} has unknown needs: {missing!r}"
            )

    if not reducers:
        raise TaskCenterInvariantViolation("Plan must contain at least one reducer")

    _assert_lane_shape(generators, reducers)
    _assert_acyclic(by_needs)

    order = _topo_order(by_needs)
    rank = {local_id: i for i, local_id in enumerate(order)}
    ordered_gen = tuple(sorted(generators, key=lambda g: rank[g.local_id]))
    ordered_red = tuple(sorted(reducers, key=lambda r: rank[r.local_id]))
    return ordered_gen, ordered_red


def _assert_lane_shape(
    generators: tuple[PlanNode, ...],
    reducers: tuple[PlanNode, ...],
) -> None:
    generator_ids = {task.local_id for task in generators}
    reducer_ids = {task.local_id for task in reducers}

    for task in generators:
        reducer_needs = tuple(dep for dep in task.needs if dep in reducer_ids)
        if reducer_needs:
            raise TaskCenterInvariantViolation(
                f"Generator task {task.local_id!r} cannot need reducer task(s): "
                f"{reducer_needs!r}"
            )

    for reducer in reducers:
        if not reducer.needs:
            raise TaskCenterInvariantViolation(
                f"Reducer task {reducer.local_id!r} must need at least one generator"
            )
        reducer_needs = tuple(dep for dep in reducer.needs if dep in reducer_ids)
        if reducer_needs:
            raise TaskCenterInvariantViolation(
                f"Reducer task {reducer.local_id!r} cannot need reducer task(s): "
                f"{reducer_needs!r}"
            )

    downstream_by_generator = {task.local_id: [] for task in generators}
    for task in generators:
        for dep in task.needs:
            if dep in generator_ids:
                downstream_by_generator[dep].append(task.local_id)
    for reducer in reducers:
        for dep in reducer.needs:
            if dep in generator_ids:
                downstream_by_generator[dep].append(reducer.local_id)

    dangling = tuple(
        local_id
        for local_id, downstream in downstream_by_generator.items()
        if not downstream
    )
    if dangling:
        raise TaskCenterInvariantViolation(
            f"Plan has generator(s) no downstream task needs: {dangling!r}"
        )


def _assert_acyclic(by_needs: dict[str, tuple[str, ...]]) -> None:
    if len(_topo_order(by_needs)) != len(by_needs):
        ordered = set(_topo_order(by_needs))
        cycle = tuple(local_id for local_id in by_needs if local_id not in ordered)
        raise TaskCenterInvariantViolation(
            f"Plan contains a dependency cycle among: {cycle!r}"
        )


def _topo_order(by_needs: dict[str, tuple[str, ...]]) -> list[str]:
    remaining = {local_id: set(needs) for local_id, needs in by_needs.items()}
    dependents: dict[str, list[str]] = {local_id: [] for local_id in by_needs}
    for local_id, needs in by_needs.items():
        for dep in needs:
            dependents[dep].append(local_id)
    ready = deque(local_id for local_id, needs in remaining.items() if not needs)
    order: list[str] = []
    while ready:
        local_id = ready.popleft()
        order.append(local_id)
        for dependent in dependents[local_id]:
            remaining[dependent].discard(local_id)
            if not remaining[dependent]:
                ready.append(dependent)
    return order


def _task_statuses_by_id(
    task_records: list[TaskRow],
) -> dict[str, TaskStatus]:
    return {task["task_id"]: TaskStatus(task["status"]) for task in task_records}


def ready_pending_plan_ids(task_records: list[TaskRow]) -> tuple[str, ...]:
    """Pending plan tasks whose ``needs`` are all DONE — ready to launch."""
    statuses = _task_statuses_by_id(task_records)
    _validate_persisted_needs(task_records, statuses)
    ready: list[str] = []
    for task in task_records:
        if statuses[task["task_id"]] != TaskStatus.PENDING:
            continue
        needs = tuple(task.get("needs") or ())
        if all(statuses[dep] == TaskStatus.DONE for dep in needs):
            ready.append(task["task_id"])
    return tuple(ready)


@dataclass(frozen=True, slots=True)
class DagStatus:
    all_quiescent: bool
    all_done: bool
    any_failed_or_blocked: bool


_FAILED_OR_BLOCKED = (TaskStatus.FAILED, TaskStatus.BLOCKED)


def _validate_persisted_needs(
    task_records: list[TaskRow],
    statuses: dict[str, TaskStatus],
) -> None:
    for task in task_records:
        missing = [dep for dep in task.get("needs") or () if dep not in statuses]
        if missing:
            raise TaskCenterInvariantViolation(
                f"Plan task {task['task_id']!r} has unknown persisted needs: {missing!r}"
            )


def _unreachable_pending_ids(
    task_records: list[TaskRow],
    statuses: dict[str, TaskStatus],
) -> frozenset[str]:
    """Pending tasks that cannot run because an upstream task failed or blocked."""
    by_id = {task["task_id"]: task for task in task_records}
    visiting: set[str] = set()
    memo: dict[str, bool] = {}

    def is_unreachable(task_id: str) -> bool:
        if task_id in memo:
            return memo[task_id]
        if task_id in visiting:
            raise TaskCenterInvariantViolation(
                f"Plan task dependency cycle reached persisted task {task_id!r}"
            )
        if statuses[task_id] != TaskStatus.PENDING:
            memo[task_id] = False
            return False

        visiting.add(task_id)
        try:
            for dep_id in by_id[task_id].get("needs") or ():
                dep_status = statuses[dep_id]
                if dep_status in _FAILED_OR_BLOCKED:
                    memo[task_id] = True
                    return True
                if dep_status == TaskStatus.PENDING and is_unreachable(dep_id):
                    memo[task_id] = True
                    return True
            memo[task_id] = False
            return False
        finally:
            visiting.remove(task_id)

    return frozenset(
        task_id
        for task_id, status in statuses.items()
        if status == TaskStatus.PENDING and is_unreachable(task_id)
    )


def dag_status(task_records: list[TaskRow]) -> DagStatus:
    """Single-pass summary of plan-task statuses for DAG dispatch.

    A pending task whose dependency chain contains a FAILED or BLOCKED task is
    quiescent but not done: it is not-started work that can never become ready
    in this attempt.
    """
    status_map = _task_statuses_by_id(task_records)
    _validate_persisted_needs(task_records, status_map)
    unreachable_pending = _unreachable_pending_ids(task_records, status_map)
    statuses = status_map.values()
    return DagStatus(
        all_quiescent=all(
            status in TERMINAL_GENERATOR_STATUSES
            or (status == TaskStatus.PENDING and task_id in unreachable_pending)
            for task_id, status in status_map.items()
        ),
        all_done=all(s == TaskStatus.DONE for s in statuses),
        any_failed_or_blocked=any(s in _FAILED_OR_BLOCKED for s in statuses),
    )

File: agent-core/crates/eos-workflow/src/attempt/plan_dag.rs
use std::collections::{BTreeMap, BTreeSet, HashMap, HashSet, VecDeque};

use eos_types::{
    AgentName, AgentRegistry, AgentType, GeneratorId, PlannerPlan, Task, TaskId, TaskStatus,
};

use crate::{Result, WorkflowError};

/// Closed scheduler resolution for a persisted plan DAG.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DagResolution {
    /// More work may still be runnable or in flight.
    Running,
    /// Every persisted plan task is done.
    Passed,
    /// The DAG is quiescent because at least one task failed or blocked.
    FailedOrBlocked,
}

/// Pending plan tasks whose needs are all done.
pub fn ready_pending_plan_ids(tasks: &[Task]) -> Result<Vec<TaskId>> {
    let statuses = statuses_by_id(tasks);
    validate_persisted_needs(tasks, &statuses)?;
    Ok(tasks
        .iter()
        .filter(|task| task.status == TaskStatus::Pending)
        .filter(|task| {
            task.needs
                .iter()
                .all(|dep| statuses.get(dep).is_some_and(|s| *s == TaskStatus::Done))
        })
        .map(|task| task.id.clone())
        .collect())
}

pub(crate) fn dag_resolution(tasks: &[Task]) -> Result<DagResolution> {
    let statuses = statuses_by_id(tasks);
    validate_persisted_needs(tasks, &statuses)?;
    let unreachable = unreachable_pending_ids(tasks, &statuses)?;
    let all_done = statuses.values().all(|s| *s == TaskStatus::Done);
    if all_done {
        return Ok(DagResolution::Passed);
    }
    let all_quiescent = statuses.iter().all(|(task_id, status)| {
        status.is_terminal_generator()
            || (*status == TaskStatus::Pending && unreachable.contains(task_id))
    });
    let any_failed_or_blocked = statuses.values().any(|s| {
        matches!(
            s,
            TaskStatus::Failed | TaskStatus::Blocked | TaskStatus::Cancelled
        )
    });
    if all_quiescent && any_failed_or_blocked {
        Ok(DagResolution::FailedOrBlocked)
    } else {
        Ok(DagResolution::Running)
    }
}

fn statuses_by_id(tasks: &[Task]) -> HashMap<TaskId, TaskStatus> {
    tasks
        .iter()
        .map(|task| (task.id.clone(), task.status))
        .collect()
}

fn validate_persisted_needs(tasks: &[Task], statuses: &HashMap<TaskId, TaskStatus>) -> Result<()> {
    for task in tasks {
        let missing: Vec<String> = task
            .needs
            .iter()
            .filter(|dep| !statuses.contains_key(*dep))
            .map(ToString::to_string)
            .collect();
        if !missing.is_empty() {
            return Err(WorkflowError::invariant(format!(
                "plan task {:?} has unknown persisted needs: {missing:?}",
                task.id.as_str()
            )));
        }
    }
    Ok(())
}

fn unreachable_pending_ids(
    tasks: &[Task],
    statuses: &HashMap<TaskId, TaskStatus>,
) -> Result<HashSet<TaskId>> {
    let by_id: HashMap<TaskId, &Task> = tasks.iter().map(|task| (task.id.clone(), task)).collect();
    let mut visiting = HashSet::new();
    let mut memo = HashMap::new();
    let mut unreachable = HashSet::new();
    for (task_id, status) in statuses {
        if *status == TaskStatus::Pending
            && is_unreachable(task_id, statuses, &by_id, &mut visiting, &mut memo)?
        {
            unreachable.insert(task_id.clone());
        }
    }
    Ok(unreachable)
}

fn is_unreachable(
    task_id: &TaskId,
    statuses: &HashMap<TaskId, TaskStatus>,
    by_id: &HashMap<TaskId, &Task>,
    visiting: &mut HashSet<TaskId>,
    memo: &mut HashMap<TaskId, bool>,
) -> Result<bool> {
    if let Some(value) = memo.get(task_id) {
        return Ok(*value);
    }
    if !visiting.insert(task_id.clone()) {
        return Err(WorkflowError::invariant(format!(
            "plan task dependency cycle reached persisted task {:?}",
            task_id.as_str()
        )));
    }
    let status = statuses
        .get(task_id)
        .copied()
        .ok_or_else(|| WorkflowError::not_found("task", task_id.as_str()))?;
    if status != TaskStatus::Pending {
        visiting.remove(task_id);
        memo.insert(task_id.clone(), false);
        return Ok(false);
    }
    let task = by_id
        .get(task_id)
        .ok_or_else(|| WorkflowError::not_found("task", task_id.as_str()))?;
    for dep in &task.needs {
        let dep_status = statuses
            .get(dep)
            .copied()
            .ok_or_else(|| WorkflowError::not_found("task", dep.as_str()))?;
        if matches!(
            dep_status,
            TaskStatus::Failed | TaskStatus::Blocked | TaskStatus::Cancelled
        ) || (dep_status == TaskStatus::Pending
            && is_unreachable(dep, statuses, by_id, visiting, memo)?)
        {
            visiting.remove(task_id);
            memo.insert(task_id.clone(), true);
            return Ok(true);
        }
    }
    visiting.remove(task_id);
    memo.insert(task_id.clone(), false);
    Ok(false)
}

pub(crate) fn validate_plan_shape(plan: &PlannerPlan) -> Result<()> {
    if plan.reducers.is_empty() {
        return Err(WorkflowError::invariant(
            "plan must contain at least one reducer",
        ));
    }
    let mut generator_ids = BTreeSet::new();
    for task in &plan.tasks {
        if !generator_ids.insert(&task.generator_id) {
            return Err(WorkflowError::invariant(format!(
                "plan contains duplicate generator id {:?}",
                task.generator_id
            )));
        }
    }

    let mut reducer_ids = BTreeSet::new();
    for reducer in &plan.reducers {
        if !reducer_ids.insert(&reducer.reducer_id) {
            return Err(WorkflowError::invariant(format!(
                "plan contains duplicate reducer id {:?}",
                reducer.reducer_id
            )));
        }
    }

    for task in &plan.tasks {
        for need in &task.needs {
            if !generator_ids.contains(need) {
                return Err(WorkflowError::invariant(format!(
                    "generator task {:?} has unknown generator needs: {:?}",
                    task.generator_id, need
                )));
            }
        }
    }
    let mut downstream_by_generator: BTreeMap<&GeneratorId, Vec<&str>> =
        generator_ids.iter().map(|id| (*id, Vec::new())).collect();
    for task in &plan.tasks {
        for need in &task.needs {
            downstream_by_generator
                .get_mut(need)
                .expect("generator needs were validated above")
                .push(task.generator_id.as_str());
        }
    }
    for reducer in &plan.reducers {
        if reducer.needs.is_empty() {
            return Err(WorkflowError::invariant(format!(
                "reducer task {:?} must need at least one generator",
                reducer.reducer_id
            )));
        }
        for need in &reducer.needs {
            if !generator_ids.contains(need) {
                return Err(WorkflowError::invariant(format!(
                    "reducer task {:?} has unknown generator needs: {:?}",
                    reducer.reducer_id, need
                )));
            }
            downstream_by_generator
                .get_mut(need)
                .expect("reducer needs were validated above")
                .push(reducer.reducer_id.as_str());
        }
    }
    let dangling: Vec<&str> = downstream_by_generator
        .iter()
        .filter_map(|(id, downstream)| downstream.is_empty().then_some(id.as_str()))
        .collect();
    if !dangling.is_empty() {
        return Err(WorkflowError::invariant(format!(
            "plan has generator(s) no downstream task needs: {dangling:?}"
        )));
    }
    assert_acyclic(plan)
}

/// Validate every plan agent before any task row is written: each generator
/// is bound to a registered workflow-launchable profile and has a task spec,
/// and the fixed `reducer` profile is registered. Runs after the pure shape
/// checks and before materialization so a rejected plan leaves no orphan rows.
pub(crate) fn validate_plan_agents(plan: &PlannerPlan, registry: &AgentRegistry) -> Result<()> {
    for task in &plan.tasks {
        let agent_name = AgentName::new(task.agent_name.clone())?;
        let agent_def = registry.get(&agent_name).ok_or_else(|| {
            WorkflowError::AgentDefinition(format!(
                "agent definition {:?} is not registered",
                task.agent_name
            ))
        })?;
        // D6: a generator task must be bound to an agent-type profile. The
        // generator role itself is task lineage, not profile metadata.
        if agent_def.agent_type != AgentType::Agent {
            return Err(WorkflowError::invariant(format!(
                "generator task {:?} is bound to agent {:?} with type {:?}, expected agent",
                task.generator_id, task.agent_name, agent_def.agent_type
            )));
        }
        if !plan.task_specs.contains_key(&task.generator_id) {
            return Err(WorkflowError::not_found(
                "task spec",
                task.generator_id.as_str(),
            ));
        }
    }
    let reducer_name = AgentName::new("reducer")?;
    let reducer = registry.get(&reducer_name).ok_or_else(|| {
        WorkflowError::AgentDefinition("agent definition \"reducer\" is not registered".to_owned())
    })?;
    if reducer.agent_type != AgentType::Agent {
        return Err(WorkflowError::invariant(format!(
            "reducer profile has type {:?}, expected agent",
            reducer.agent_type
        )));
    }
    Ok(())
}

fn assert_acyclic(plan: &PlannerPlan) -> Result<()> {
    let mut by_needs: BTreeMap<&str, Vec<&str>> = BTreeMap::new();
    for task in &plan.tasks {
        by_needs.insert(
            task.generator_id.as_str(),
            task.needs.iter().map(GeneratorId::as_str).collect(),
        );
    }
    let mut remaining = by_needs
        .iter()
        .map(|(id, needs)| (*id, needs.iter().copied().collect::<BTreeSet<_>>()))
        .collect::<BTreeMap<_, _>>();
    let mut dependents: BTreeMap<&str, Vec<&str>> =
        by_needs.keys().map(|id| (*id, Vec::new())).collect();
    for (id, needs) in &by_needs {
        for need in needs {
            if let Some(entries) = dependents.get_mut(need) {
                entries.push(id);
            }
        }
    }
    let mut ready = remaining
        .iter()
        .filter_map(|(id, needs)| needs.is_empty().then_some(*id))
        .collect::<VecDeque<_>>();
    let mut order = Vec::new();
    while let Some(id) = ready.pop_front() {
        order.push(id);
        for dependent in dependents.get(id).into_iter().flatten() {
            if let Some(needs) = remaining.get_mut(dependent) {
                needs.remove(id);
                if needs.is_empty() {
                    ready.push_back(dependent);
                }
            }
        }
    }
    if order.len() != by_needs.len() {
        let ordered = order.into_iter().collect::<BTreeSet<_>>();
        let cycle = by_needs
            .keys()
            .filter(|id| !ordered.contains(**id))
            .copied()
            .collect::<Vec<_>>();
        return Err(WorkflowError::invariant(format!(
            "plan contains a dependency cycle among: {cycle:?}"
        )));
    }
    Ok(())
}

#[cfg(test)]
#[path = "../../tests/attempt/plan_dag/mod.rs"]
mod tests;
